### ISSUE TITLE
android.permission.QUERY_ALL_PACKAGES

### ISSUE THREAD

**Issue by IzzySoft** [vpnhood/VpnHood#449]
My freshly enhanced "security scanner" (your app is available [in my repo](https://apt.izzysoft.de/packages/com.vpnhood.client.android/) – so if you wish, be welcome to [pick a badge](https://gitlab.com/IzzyOnDroid/repo/#if-my-app-is-available-here-do-you-have-a-badge-i-can-use-to-link-to-it) pointing to that link e.g. from next to the ones of Play and Github) just reported:

    ! repo/com.vpnhood.client.android_454.apk declares risky permissions: android.permission.QUERY_ALL_PACKAGES

May I ask what for this permission is required? Does a VPN client really need to list all installed apps?

PS: I already added the other findings to the allow-list of VpnHood as their need is rather clear:

```
! repo/com.vpnhood.client.android_454.apk declares flags: usesCleartextTraffic
! repo/com.vpnhood.client.android_454.apk declares intent-filters: android.net.VpnService
```

A VPN client needs that intent, and it also needs to route traffic that's not secured by HTTPS, so no questions on those two :wink:

**Comment by trudyhood** [comment]
Hi, Thank you for asking this.

We need to use cleartext traffic to our internal web server for our SPA (the UI), which is hosted on localhost, so it cannot be configured for HTTPS, which is not needed either. All VpnHood outbound connections are strongly encrypted.

We require the android.permission.QUERY_ALL_PACKAGES permission to enable our VPN app to split tunnel traffic for specific apps, giving users the flexibility to choose which apps use the tunnel. We already created a video for Google Play so that I can share it.
https://drive.google.com/file/d/1Ar27T8_Ejiplr1KgFp2wNN7pfG2kBRIV/view?usp=sharing

Ideally, Android would manage app splitting for VPNs natively, eliminating the need for us to implement this feature.

Let me know if you have any more questions.

**Comment by IzzySoft** [comment]
Thanks a lot, @trudyhood – that explains it well (though I had to skip the video as I have issues with GDrive). Added all that to your app's allow-list then with your explanations (this will at some future point become visible in its own section on the app details pages, for transparency).

**Comment by IzzySoft** [comment]
@trudyhood sorry to bother you once more, but why was the Google Play Core library added with the last release? That is proprietary and reaches out to Google servers, which makes users and their actions traceable, and thus is not a good thing to have in a privacy-focused app. In my repo, I now have to raise the `NonFreeComp` (includes non-free/proprietary component) and `NonFreeNet` (non-free network used) anti-features which, with a privacy app, make it "less desirable".

My guess is it is related to "Feature: Android: Prompt for the update by Google Play if possible" from the changelog. That might be OK for those having installed your app from PlayStore, but will certainly annoy those who rather avoid PlayStore/Google and rather get your app via my repo. Maybe as a compromise, to serve both, there could be a "libre build flavor" without that dependency (with its APK provided next to the other at releases here and maybe using a file name like `VpnHoodClient-android-libre.apk`, if it's not too much trouble, so those flags can be removed when I switch to that APK (I could then configure my updater to only fetch APKs matching the RegEx `/libre/i`)?

(PS: between the lines you certainly saw it already, there's an updater running once a day to check for and fetch updates to my repo from your releases here at Github. Those having installed the app from there will then receive update notifications from their [F-Droid client](https://android.izzysoft.de/applists/category/named/apps_markets#group_1181), same way GooglePlay does. Oh, this also means you could [pick a badge](https://gitlab.com/IzzyOnDroid/repo/#if-my-app-is-available-here-do-you-have-a-badge-i-can-use-to-link-to-it) to place next to the one of Google and [link to your app in my repo](https://apt.izzysoft.de/packages/com.vpnhood.client.android/))

**Comment by trudyhood** [comment]
We already have a Google Free app with a different package name, "com.vpnhood.client.android.web," but we haven't uploaded it officially. We can upload it if you are interested. This package is completely isolated and installed with the other one side by side, and the difference is the background icon color.
I think we need to do a little work on it to exclude the unnecessary Google libraries that it doesn't use.

**Comment by IzzySoft** [comment]
Thanks Trudy! Whichever variant you prefer. If it goes by a different packageName (like the `.web`) it will basically be a new app I'd add and, after a decent overlap with a hint in the current one that a replace is in progress, would remove the one present in my repo now. If it were a separate build flavor of the current app, going by the same packageName as the one listed, it would just be "outaging previous APKs" and the id/link would stay unchanged.

**Comment by trudyhood** [comment]
we had many problems by having same package name with two different signing. if we going to have separate build we the same package name we need to use Google signature which is not comply by Google policy. 

That's the bad of Google, they have many policies and restrictions :(

**Comment by trudyhood** [comment]
Here is the Google free version. I tried to exclude Google Libraries but didn't analyze the output. I hope everything is ok.
https://github.com/vpnhood/VpnHood/releases/download/v3.3.454/VpnHoodClient-android-web.apk

**Comment by trudyhood** [comment]
Link to the latest version:
https://github.com/vpnhood/VpnHood/releases/latest/download/VpnHoodClient-android-web.apk

**Comment by IzzySoft** [comment]
Thanks Trudy! And here's the output of my scanner for it:

```
Libraries detected:
-------------------
* Google Play Core (/com/google/android/play/core): Utility, Proprietary; NonFreeNet,NonFreeComp
* Java.Interop (/com/xamarin/java_interop): Utility, MIT
* Mono for Android (/mono/android): Utility, MIT
* Xamarin (/xamarin): Development Framework, MIT

Offending libs:
---------------
* Google Play Core (/com/google/android/play/core): NonFreeNet,NonFreeComp

1 offenders.

Dangerous flags:
----------------
* usesCleartextTraffic

Dangerous Filters:
------------------
android.net.VpnService

Permissions:
------------
* android.permission.INTERNET
* android.permission.QUERY_ALL_PACKAGES
* android.permission.POST_NOTIFICATIONS
* android.permission.FOREGROUND_SERVICE
* android.permission.FOREGROUND_SERVICE_SYSTEM_EXEMPTED
* android.permission.START_FOREGROUND_SERVICES_FROM_BACKGROUND
```

Looks like it still has the Play Core lib inside. All else was already on the allow-list for the other variant. So from this point, I see no difference (Play Core is also the only "offender" listed there).

**Comment by trudyhood** [comment]
Oh my bad. I will recheck it. 
Do you know any online tool so I can check it before publishing?

**Comment by trudyhood** [comment]
The issue resided in a weird behavior of msbuild. I replaced it with dotnet build - gubbled up 3 hours for me to find the root of issue :( 

I checked it via https://www.sisik.eu/apk-dump and couldn't find any trace of Google Play anymore.
Please check again:
https://github.com/vpnhood/VpnHood/releases/latest/download/VpnHoodClient-android-web.apk

I have also configured the network_security_config and limited use of cleartext traffic to localhost; let me know about it, too.

**Comment by IzzySoft** [comment]
Looks like a full success:

```
Libraries detected:
-------------------
* Java.Interop (/com/xamarin/java_interop): Utility, MIT
* Mono for Android (/mono/android): Utility, MIT
* Xamarin (/xamarin): Development Framework, MIT

No offending libs found.


Dangerous Filters:
------------------
android.net.VpnService

Permissions:
------------
* android.permission.INTERNET
* android.permission.QUERY_ALL_PACKAGES
* android.permission.POST_NOTIFICATIONS
* android.permission.FOREGROUND_SERVICE
* android.permission.FOREGROUND_SERVICE_SYSTEM_EXEMPTED
* android.permission.START_FOREGROUND_SERVICES_FROM_BACKGROUND
```

The lib is gone, and the flag as well. I didn't expect the latter. Still working as it should, with "local cleartext"?

**Comment by trudyhood** [comment]
> The lib is gone, and the flag as well. I didn't expect the latter. Still working as it should, with "local cleartext"?

It is still necessary to serve SPA, but I have limited permissions to localhost by configuring network_security_config.xml.

[VpnHood.Client.App.Android/Resources/xml/network_security_config.xml]
(https://github.com/vpnhood/VpnHood/blob/development/VpnHood.Client.App.Android/Resources/xml/network_security_config.xml)

The package ID for this distribution must be com.vpnhood.client.android.web

Thank you for analyzing our repository.

**Comment by IzzySoft** [comment]
Gladly done, Trudy! It was a joy with your showing real interest, picking up and finally tackling the culprit(s)!

So do you want me to switch to the `.web` variant then in my repo? Existing users would need to switch manually then.

**Comment by trudyhood** [comment]
I appreciate your compliment. 
Currently, we focus on distributing the Google Play version. However, we don't know about your clients. I am sure you know more about your client's preferences. The .web version is for those who are afraid of Google services. The choice is yours.

**Comment by IzzySoft** [comment]
@trudyhood today's update failed as there was no APK with the package name used until now. Shall I move over to the `.web` one then? It would basically work like this:

* I establish the `.web` one as new app, while at the same time disable updates for the previous one and add a note to its description that folks currently using that should (manually) switch over
* this constellation is kept for 2..4 weeks
* at the end of the "migration period", the "old app" is removed from my repo

Does that suit you?

**Comment by trudyhood** [comment]
Hi, Nice to hear from you
The APK will be updated soon, and we are waiting for Google review.
I am still unsure which APK we should put in your repository to suit the client's needs. Your decision and experience will be appreciated.

**Comment by IzzySoft** [comment]
TBH, with the privacy focus I have plus the added FOSS one I'd switch to the web variant with my repo. This would rob folks the chance for a "simple switch" from Play to my repo, though. But with an app that I'd expect to have privacy at its core (we're talking about a VPN here, and the P in VPN Probably (oops) is not meant for "Pay" or "Problem" but for … oh: "Private"? Matches, huh? :see_no_evil:) I'd say that's a Price (oof) to Pay (eh, enough games now :rofl:).

So I'd suggest to do that switch here, if you agree. It would also mean the AntiFeatures (`NonFreeComp` and `NonFreeNet`) could most likely be removed, which to several folks using my repo might currently be a blocker to use your app. The 200 kB space saved are a nice addition to that but not really an argument.

So if you'd say "yes, go for it" I'd initiate the switch immediately, so the web variant would show up with the next sync around 6 pm UTC **in addition** to the current one (with the current one not updated anymore – or updated one last time once it becomes available here so whoever applies updates manually will get a better chance to see the notification for switching over).

Those who want the Google Services in are probably fine getting the current variant by other means (PlayStore app, AuroraStore, Obtainium, Raccoon, whatever) – and those who are not will be happy not to have to. I don't think anyone currently using VpnHood would drop it because of that, but some of those having refused it because of the "tainting" might want to give it a try with the then fully FOSS one.

**TL;DR:** I'd suggest making the switch now, and would strongly prefer it. But as it's your app, I'd leave the decision to you.

**Comment by IzzySoft** [comment]
@trudyhood that update arrived today, but I had to "kill" it and disable updates for now: the APK triggers a rare bug with androguard, so index build fails unless I move it out of the repo :cry:

Please let me know if it's OK with you to switch to the `.web` variant here to get the updates flowing.

**Comment by IzzySoft** [comment]
@trudyhood alert! the [VpnHoodClient-android-web.apk](https://github.com/vpnhood/VpnHood/releases/download/v4.3.485/VpnHoodClient-android-web.apk) is signed using a debug certificate, so I cannot take it. Please replace it with a properly signed version ASAP.

Btw: I've fixed the bug with androguard (patching it manually), so the current release will be available in a few minutes (had to run a manual update for another reason, so I decided to push it out manually). But to keep the mess at bay, please let us perform the switch ASAP. Thanks!

**Comment by trudyhood** [comment]
Let me work on that

**Comment by IzzySoft** [comment]
Just wondering what the status is here? Also wondering that the Google dependencies are gone from the "regular" APK, so maybe a switch is no longer needed?

**Comment by trudyhood** [comment]
Hello again! First, thank you for reporting the debug signature issue to me. I just fixed it; please let me know if there is something wrong with the updated APK. 
Regarding Google Dependencies, there is a Google Play Core dependency for the update prompt. However, I tested it in an emulator with a Google API emulator and found out that it works without any problem.

**Comment by IzzySoft** [comment]
Thanks @trudyhood! So we stick with switching to the `.web` version in my repo (and future releases of that APK will be signed with the same key as the current one, which I hereby can confirm is fine)?

**Comment by trudyhood** [comment]
It's okay. However, if you find that the Google Play version can work on any device without relying on Google Play dependencies, you can continue using it.

**Comment by IzzySoft** [comment]
I cannot try as I don't use VPN on my Android devices, sorry. But if you prefer I can of course keep that one in my repo – with the AntiFeatures applied. It's your app, and hence your decision (I must have missed something, the Google Play Core library is still in – is that really that important?)

**Comment by trudyhood** [comment]
Thank you for your information. The Google Play Core library is required in our app for the Google Play update prompt, and it is crucial for us as the app continuously updates and will stop working if the user doesn't update it. So, we need that until the app reaches an LTS version. We are okay with the AntiFeatures label.

Has VpnHood included includeInApk and includeInBundle? We use .net for Android (formerly Xamarin), and no such settings exist, or at least I couldn't find them.

**Comment by IzzySoft** [comment]
> The Google Play Core library is required in our app for the Google Play update prompt

which doesn't work if no Google Services (and not PlayStore App) are available on the device – which is the case with many of those using my repo. Further, even if they were, those people would not be happy to get their updates via PlayStore, which is why those are using F-Droid's and my repo. Their [F-Droid client](https://android.izzysoft.de/applists/category/named/apps_markets#group_1181)s will take care for the updates.

> We are okay with the AntiFeatures label.

OK, then let's leave things as they are for now. Once you reach LTS, you'll probably drop this library?

> Has VpnHood included includeInApk and includeInBundle? We use .net for Android (formerly Xamarin), and no such settings exist, or at least I couldn't find them.

The default is to include those – so you don't need the explicit setting if you want that in, only if you want that out. As for my repo, it's sufficient to exclude it for the APK, so you can keep it in for the bundle (to e.g. upload to PlayStore) if you wish. For my repo I'd prefer to **not** have it – for the outlined reason. It's no show-stopper for now, and does not even raise an AntiFeature (at least not yet) – but it shows up with the package details to be transparent for those concerned:

![image](https://github.com/vpnhood/VpnHood/assets/6781438/85a439be-2cc4-4766-952a-b5a603392d28)

The "info icon" links to [a description](https://apt.izzysoft.de/fdroid/index/info#signingblock).

**Comment by trudyhood** [comment]
We need to prompt users to update the app regularly because the old release might stop working due to connectivity issues. We don't maintain much backward compatibility, so it's important that users keep up with the updates.

I couldn't find the signing block in Xamarin, and I think it's OK for Google Play distribution. However, web distribution might be a better option for users who can't use or don't want to use Google.

**Comment by IzzySoft** [comment]
> I couldn't find the signing block in Xamarin

I've no clue about Xamarin, sorry. Thought maybe like Flutter, it generates the proper tree for Android and then uses the `build.gradle` from there – but that's totally out of my scope unfortunately. Signing [happens here](https://github.com/vpnhood/VpnHood/blob/main/Pub/Core/PublishAndroidApp.ps1#L56), but I don't know what `dotnet build` uses for signing. And I also missed that what remains here is not the `DEPENDENY_INFO_BLOCK`, but rather the `SOURCE_STAMP_V2_BLOCK`, which is less critical as it's at least not encrypted (see [SigningBlock checks](http://apt.izzysoft.de/fdroid/index/info#signingblock) for some details).

> I think it's OK for Google Play distribution

Yeah, they may even want this, so you can definitely leave it in the AAB.

> web distribution might be a better option for users who can't use or don't want to use Google.

An F-Droid client is much better for that. Which is why your app is in my repo. So like the PlayStore app, people receive update notifications from their [F-Droid client](https://android.izzysoft.de/applists/category/named/apps_markets#group_1181).

**Comment by IzzySoft** [comment]
@trudyhood I wonder if you let Google do the signing? I see `SOURCE_STAMP_V2_BLOCK` is still there. So far I only saw that with APKs signed by Google. Could AFAIK also happen if you create your APKs from app bundles (AABs).

**Comment by trudyhood** [comment]
Hi,

This is signed and downloaded from Google Play:
https://github.com/vpnhood/VpnHood/releases/download/v5.0.635/VpnHoodClient-android.apk

This one is Google Free but has a different package ID
https://github.com/vpnhood/VpnHood/releases/download/v5.0.635/VpnHoodClient-android-web.apk

Which one does have SOURCE_STAMP_V2_BLOCK?

**Comment by IzzySoft** [comment]
`com.vpnhood.client.android` – so that would be the former, matches. I vaguely remember at one point we discussed to switch to `com.vpnhood.client.android.web`, but for some reason did not (ah, scrolling up I see we discussed this [here a year ago](https://github.com/vpnhood/VpnHood/issues/449#issuecomment-1901168255)).

So the `.web` version comes without the Google Play Core lib and without the `SOURCE_STAMP_V2_BLOCK` – and that's the major differences our scanners report. Switching here (at IzzyOnDroid) to the `.web` version would mean who already uses VpnHood from here must switch to the other variant (i.e. uninstall the Google variant and install the Web variant). Can settings etc. be easily transferred? Then we should really consider it: the Web variant also would not need the NonFreeNet and NonFreeDep anti-features.

**Comment by trudyhood** [comment]
We have tried to keep the .web version with minimum dependencies and completely Google-free. We changed the packaged ID (same as what telegram does) as they can be installed side by side without conflict; its icon background is slightly different, so distinguish the distribution.  There are a few settings; users only need to add their server keys. Let me know if there is any anti-feature in this distribution.

**Comment by IzzySoft** [comment]
As I wrote (apologies if I was not clear there): I see no reason to have any anti-feature on the `.web` variant. No "offending libs", no blobs, nothing:

```
$ iod shell scanapk VpnHoodClient-android-web.apk

Libraries detected:
-------------------
* Mono for Android (/mono/android): Utility, MIT
* Xamarin (/xamarin): Development Framework, MIT

No offending libs found.


Dangerous Filters:
------------------
android.net.VpnService

Permissions:
------------
* android.permission.INTERNET
* android.permission.QUERY_ALL_PACKAGES
* android.permission.POST_NOTIFICATIONS
* android.permission.FOREGROUND_SERVICE
* android.permission.FOREGROUND_SERVICE_SYSTEM_EXEMPTED
* android.permission.START_FOREGROUND_SERVICES_FROM_BACKGROUND

Signing:
--------
package: name='com.vpnhood.client.android.web' versionCode='635' versionName='5.0.635' platformBuildVersionName='15' platformBuildVersionCode='35' compileSdkVersion='35' compileSdkVersionCodename='15'
Verifies
Verified using v1 scheme (JAR signing): true
Verified using v2 scheme (APK Signature Scheme v2): true
Verified using v3 scheme (APK Signature Scheme v3): true
Verified using v4 scheme (APK Signature Scheme v4): false
Verified for SourceStamp: false
Number of signers: 1
Signer #1 certificate DN: CN="VpnHood, O=OmegaHood LLC, S=CA, C=US"
Signer #1 certificate SHA-256 digest: 07491d8a6547058071691e5d5dffc1a8b2c2584e11f27857f49b58d0a6bb7bf4
Signer #1 certificate SHA-1 digest: 355c1b88ad89d8b1c842decfca758410e8cfaa72
Signer #1 certificate MD5 digest: 5cd2c44c13247d3ef4a8a411374b6bf1
Signer #1 key algorithm: RSA
Signer #1 key size (bits): 2048
Signer #1 public key SHA-256 digest: 49cb9a8e874602297c44badd9dca35f879115ef329966df7001c3d97390ae3b4
Signer #1 public key SHA-1 digest: c319300e1556137c5ea8ec24421e63e74443562c
Signer #1 public key MD5 digest: 774fe69cc88bedab8f3bce1754bd5575
```

That looks totally clean for me. For considering a switch to this variant at IzzyOnDroid, let me raise the following questions/remarks:

1. Is there an easy way to transfer settings from the (installed) "Google" variant to the Web variant (installed in parallel, or after removal of the former)?
2. Would it be OK for you that those getting your app via IzzyOnDroid would have to switch, as the "Google" variant would no longer be kept here then (after a decent overlap of course – say, 1..2 months)?
3. Both of the above would then also apply to anyone wanting to "switch over" from a PlayStore install to IzzyOnDroid; so with the Web variant here, it's no longer just replacing the source.

Those are the things to be considered. If all sounds good to you, we can do the switch:

1. I'd establish the Web variant at IzzyOnDroid
2. I'd include a hint with the Google variant, pointing out IzzyOnDroid will switch to the Web variant (linking to it so people can find it easier)
3. I'd mark the Google variant for removal (setting a date)
4. Once the removal date is due, I can optionally send you a ping first to confirm – then remove the Google variant
5. switch completed.

Just let me know. Meanwhile, some numbers, as we now have download stats:

* downloads in the last 5 weeks (since stats started): 2835
* day-by day (tap to open):

<details><summary>Daily stats for com.vpnhood.client.android</summary><pre>
2024-12-08: 16
2024-12-09: 19
2024-12-10: 17
2024-12-11: 21
2024-12-12: 13
2024-12-13: 14
2024-12-14: 13
2024-12-15: 2
2024-12-15: 16
2024-12-16: 18
2024-12-17: 114
2024-12-18: 323
2024-12-19: 187
2024-12-20: 129
2024-12-21: 117
2024-12-22: 76
2024-12-23: 80
2024-12-24: 70
2024-12-25: 51
2024-12-26: 60
2024-12-27: 64
2024-12-28: 202
2024-12-29: 1
2024-12-29: 356
2024-12-30: 153
2024-12-31: 86
2025-01-01: 89
2025-01-02: 83
2025-01-03: 76
2025-01-04: 61
2025-01-05: 52
2025-01-06: 45
2025-01-07: 44
2025-01-08: 38
2025-01-09: 44
2025-01-10: 38
2025-01-11: 47</pre></details>

**Comment by IzzySoft** [comment]
PS: you can [find the statistics here](https://codeberg.org/IzzyOnDroid/iodstats/) if you want to look up future ones 😉

**Comment by trudyhood** [comment]
> Is there an easy way to transfer settings from the (installed) "Google" variant to the Web variant (installed in parallel, or after removal of the former)?

It requires much work to be done. The user has keys; there are no other settings, and they can easily migrate themself. I don't think it is a serious issue for them. 

> Those are the things to be considered. If all sounds good to you, we can do the switch:

It all sounds good.

> Just let me know. Meanwhile, some numbers, as we now

Nice. Is there any statistic to show how many are active or uninstalled after installation?

We have another app, too. Does IzzyOnDroid support commercial apps and open-source apps, too? Such as having AD or a required subscription?

**Comment by IzzySoft** [comment]
> It all sounds good.

So shall we do it for real this time? 😉

> Is there any statistic to show how many are active or uninstalled after installation?

No. We do not track anything. The download stats are just "extracted" from our web server logs when the latter are rotated, only date + app + count. We don't check which apps are installed on devices, or anything else. No accounts needed with IzzyOnDroid.

> Does IzzyOnDroid support commercial apps and open-source apps, too?

Only open-source apps, under a libre license (full code available, and all [4 essential freedoms of free software](https://en.wikipedia.org/wiki/Free_and_open-source_software#Four_essential_freedoms_of_Free_Software) given). If some API keys are needed for "full operation" and can be acquired separately (and then entered manually into the app), that should be made clear in the app description. If a "subscription" means a commercial/non-libre" network connection (which cannot be replaced by something self-hosted), that would require the NonFreeNet anti-feature (many apps carry that; now that I think about it, that's the only one we need to keep with the Web variant here, too, as VpnHood certainly only works with your service, and the server part is not libre I guess?).

> Such as having AD

What is AD? Advertising? Well, while not exactly a show-stopper (we have an "Ads" anti-feature), that much depends on the kind of ads. Your own, shipped from your servers? Or requiring some (closed-source) libraries from other ad networks? You might wish to take a look at the [IzzyOnDroid App Inclusion Policy](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Inclusion-Policy) for details.

**Comment by trudyhood** [comment]
> So shall we do it for real this time

Yes

> VpnHood certainly only works with your service, and the server part is not libre I guess?).

No, VpnHood is entirely free, including its server and client. Anyone can install its server and use the client or fork the project, building their app or their server. No part needs a commercial license.

We have another app called VpnHood! CONNECT, which is a fork of this app and works exclusively with our official server. In the Google release version, ads are displayed using Google AdMob. For the web release, users need to order an access key from us. 

The difference between the two apps is that VpnHood! CLIENT (currently available on IzzyOnDroid) accepts keys from anyone who creates a server, while VpnHood! CONNECT only accepts keys from the official VpnHood server. In addition, VpnHood! CONNECT has built an official server and offers a free trial connection for a few minutes. All sources is free, same as VpnHood! CLIENT. the difference is the configuration, not the code.

**Comment by IzzySoft** [comment]
> > So shall we do it for real this time

> yes

OK, switch initiated. Would you consider per-ABI builds – or an APK coming only with the 2 ARM ABIs? I'm not aware of any x86 devices, and those native libraries taking up quite a lot of space there. An APK without x86 libs would certainly be below the 30 MB margin we usually have here per-app at IzzyOnDroid.

> No, VpnHood is entirely free, including its server and client. Anyone can install its server and use the client or fork the project, building their app or their server. No part needs a commercial license.

Wonderful! Then the Web variant will be established without any anti-features. I've set up a due-date for removing the Google variant here around 2025-03-20, and put a note on it so point those currently using it to switch over. Until the due-date, both variants will be updated here should you have new releases (usually at this point, updates for the "old variant" are disabled – but as this is a security relevant app, we'll make an exception here again, as we already did with the app size).

> We have another app called VpnHood! CONNECT

[Vpnhood.App.Connect](https://github.com/vpnhood/Vpnhood.App.Connect)? Hm… As everything this app here contains is available there, maybe we can first speak about APK size and per-ABI builds? That APK is also ~40 MB, while we have a per-app limit of 30 MB. A single-ABI (eg. arm64) certainly would be below 20 MB…

**Comment by trudyhood** [comment]
> I'm not aware of any x86 devices,

x86/64 exists in a significant number of Android TV devices, not phones. We found that for many people, 30MB or 70MB is not very important; however, in some situations, people may care about it.  We prefer a universal APK as it is easier to maintain. You may need to make an exception for the size of an app that runs on TV with x86/x64. Apps such as VPN are very important on Android TV, as they open many streaming sites. We tried to drop those binaries but then noticed many Android TV devices need them.

> [Vpnhood.App.Connect](https://github.com/vpnhood/Vpnhood.App.Connect)? Hm… As eve

Yes, the releases are here, and its repo is just for the binaries, not source code; it compiles and publishes directly from the main VpnHood repo. We have also decoupled VpnHood! SERVER release from the main repo to prevent the need to increase the version and publish to the main repo while there was no change in clients. All the source code are available in the main repo and ready to build.

https://github.com/vpnhood/Vpnhood.App.Connect/releases
